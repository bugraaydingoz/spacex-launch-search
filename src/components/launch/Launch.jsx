import React from 'react'

// Components
import { Button } from '../button'
import { Status } from '../status'

// Styles
import {
  LaunchContainer,
  Image,
  DetailsContainer,
  Heading3,
  Row,
  Column,
  FlexEnd,
  Paragraph,
  Text,
  DetailText,
  Dot,
} from './styles'

function truncate(text, length) {
  if (!text) {
    return ''
  }

  if (text.length < length) {
    return text
  }

  return text.substr(0, length) + '...'
}

function getStatus(success, upcoming) {
  if (upcoming) {
    return 'future'
  }

  if (success) {
    return 'success'
  }

  return 'failure'
}

function formatDate(utc) {
  const date = new Date(utc)
  return `${date.getDate()}.${date.getMonth()}.${date.getFullYear()}`
}

export const Launch = React.memo((props) => {
  const {
    name,
    links,
    details,
    success,
    upcoming,
    date_utc,
    rocketName,
  } = props
  const imageUrl = links.flickr.original[0] || './icons/space-ship.svg'

  function openTab(link) {
    const tab = window.open(link, '_blank')
    tab.focus()
  }

  return (
    <LaunchContainer>
      <Image src={imageUrl} alt="Launch picture" lazy />
      <DetailsContainer>
        {/* Title */}
        <Row style={{ justifyContent: 'unset' }}>
          <Heading3>{name}</Heading3>
          <Status type={getStatus(success, upcoming)} />
        </Row>

        {/* Details */}
        <Row>
          <Paragraph>{truncate(details, 80)}</Paragraph>
        </Row>
        <Row>
          <Column>
            <Details rocket={rocketName} date={formatDate(date_utc)} />
          </Column>
          <FlexEnd>
            <Button onClick={() => openTab(links.article)}>LEARN MORE</Button>
          </FlexEnd>
        </Row>
      </DetailsContainer>
    </LaunchContainer>
  )
})

function Details({ rocket, date }) {
  return (
    <>
      <DetailText>
        <Dot />
        &nbsp; Launched on &nbsp;<Text primary>{date}</Text>
      </DetailText>
      <DetailText>
        <Dot />
        &nbsp;&nbsp;<Text primary>{rocket}</Text>&nbsp;&nbsp;rocket is used
      </DetailText>
    </>
  )
}
