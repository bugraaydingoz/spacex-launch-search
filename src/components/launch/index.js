export { Launch } from './Launch'
export { LaunchList } from './LaunchList'
export { LaunchLoading } from './LaunchLoading'
