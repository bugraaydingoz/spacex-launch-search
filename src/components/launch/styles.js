import styled from 'styled-components'

export const LaunchListContainer = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0 auto 60px auto;
  width: 660px;

  @media (max-width: 768px) {
    width: 100%;
  }
`

export const LaunchContainer = styled.div`
  display: flex;
  margin-bottom: 30px;
  width: 100%;
  height: 210px;
  background-color: var(--elevation-color);

  &:last-child {
    margin-bottom: 0;
  }

  @media (max-width: 768px) {
    flex-direction: column;
    height: auto;
  }
`

export const Image = styled.img`
  display: flex;
  background-color: #1f1f1f;
  width: 220px;
  height: 100%;

  @media (max-width: 768px) {
    height: 100px;
    align-self: center;
  }
`

export const DetailsContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  max-width: 440px;
  padding: 20px 30px;
`

export const Heading3 = styled.h3`
  margin: 0 15px 20px 0;
  font-weight: bold;
  font-size: 24px;
  color: #fff;
`

export const Heading5 = styled.h5`
  margin: 0 0 10px 0;
  font-weight: normal;
  font-size: 20px;
  color: var(--secondary-text);
`

export const Paragraph = styled.p`
  margin: 0 0 40px 0;
  font-weight: normal;
  font-size: 18px;
  color: #fff;
`

export const Row = styled.div`
  display: flex;
  justify-content: space-between;
`

export const Column = styled.div`
  display: flex;
  flex-direction: column;
`

export const FlexEnd = styled.div`
  display: flex;
  align-items: flex-end;
`

export const Dot = styled.div`
  display: flex;
  width: 3px;
  height: 3px;
  border-radius: 999px;
  background-color: #fff;
`
export const Text = styled.span`
  display: inline-flex;
  font-weight: bold;
  font-size: 16px;
  color: #fff;
`

export const DetailText = styled.span`
  display: flex;
  align-items: center;
  margin: 0 0 5px 0;
  font-weight: normal;
  font-size: 16px;
  color: var(--secondary-text);
`
