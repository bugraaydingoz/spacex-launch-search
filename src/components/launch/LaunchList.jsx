import React from 'react'

// Styles
import { LaunchListContainer, Heading5 } from './styles'

export function LaunchList({ children, loading, length }) {
  const text = loading ? 'Fetching launches' : `${length} launches found`
  return (
    <LaunchListContainer>
      <Heading5>{text}</Heading5>
      {children}
    </LaunchListContainer>
  )
}
