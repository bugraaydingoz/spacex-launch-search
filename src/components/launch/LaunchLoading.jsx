import React from 'react'

// Styles
import { LaunchContainer } from './styles'

export function LaunchLoading() {
  return <LaunchContainer />
}
