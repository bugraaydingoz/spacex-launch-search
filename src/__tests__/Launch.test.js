import React from 'react'
import renderer from 'react-test-renderer'
import { Launch } from '../components/launch'

const launch = {
  name: 'test',
  details: 'test',
  success: true,
  upcoming: false,
  links: {
    flickr: {
      original: ['test'],
    },
    article: 'test',
  },
  date_utc: '2020-09-01T00:00:00.000Z',
}

test('renders correctly', () => {
  const tree = renderer.create(<Launch {...launch} />).toJSON()
  expect(tree).toMatchSnapshot()
})
