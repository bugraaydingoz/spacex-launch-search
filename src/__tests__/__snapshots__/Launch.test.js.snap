// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-AxhCb fxxzMA"
>
  <img
    alt="Launch picture"
    className="sc-AxhUy jvaykL"
    src="test"
  />
  <div
    className="sc-AxgMl kGZRVh"
  >
    <div
      className="sc-fzozJi gEVIIe"
      style={
        Object {
          "justifyContent": "unset",
        }
      }
    >
      <h3
        className="sc-AxheI ijdgJv"
      >
        test
      </h3>
      <div
        className="sc-AxirZ Linil"
        type="success"
      >
        SUCCESS
      </div>
    </div>
    <div
      className="sc-fzozJi gEVIIe"
    >
      <p
        className="sc-AxmLO gBaJEW"
      >
        test
      </p>
    </div>
    <div
      className="sc-fzozJi gEVIIe"
    >
      <div
        className="sc-fzoLsD iAaohI"
      >
        <span
          className="sc-fznKkj iVLJTV"
        >
          <div
            className="sc-fzplWN dxnSRB"
          />
            Launched on  
          <span
            className="sc-fznyAO fnMzL"
          >
            1.8.2020
          </span>
        </span>
        <span
          className="sc-fznKkj iVLJTV"
        >
          <div
            className="sc-fzplWN dxnSRB"
          />
            
          <span
            className="sc-fznyAO fnMzL"
          >
            Falcon 1
          </span>
            rocket is used
        </span>
      </div>
      <div
        className="sc-fzpans jpxutW"
      >
        <button
          className="sc-AxjAm lhlWzM"
          onClick={[Function]}
        >
          LEARN MORE
        </button>
      </div>
    </div>
  </div>
</div>
`;
